octree/basic_utils_test.go

```diff
@@ -0,0 +1,187 @@
+
+		test.That(t, basicOct.nodeType, test.ShouldResemble, LeafNodeEmpty)
+		test.That(t, basicOct.point, test.ShouldResemble, pc.PointAndData{})
+		test.That(t, basicOct.tree, test.ShouldBeNil)
+	})
+
+	t.Run("Create filled leaf node", func(t *testing.T) {
+		p := r3.Vector{X: 0, Y: 0, Z: 0}
+		d := pc.NewValueData(1.0)
+		basicOct := newLeafNodeFilled(p, d)
+
+		test.That(t, basicOct.nodeType, test.ShouldResemble, LeafNodeFilled)
+		test.That(t, basicOct.point, test.ShouldResemble, pc.PointAndData{P: p, D: d})
+		test.That(t, basicOct.tree, test.ShouldBeNil)
+	})
+
+	t.Run("Create internal node", func(t *testing.T) {
+		var children []*basicOctree
+		basicOct := newInternalNode(children)
+
+		test.That(t, basicOct.nodeType, test.ShouldResemble, InternalNode)
+		test.That(t, basicOct.point, test.ShouldResemble, pc.PointAndData{})
+		test.That(t, basicOct.tree, test.ShouldResemble, children)
+	})
+}
+
+// Tests that the splitting of a filled octree node results in seven empty leaf nodes
+// and one filled one or an empty octree node will result in eight empty children nodes.
+func TestSplitIntoOctants(t *testing.T) {
+	ctx := context.Background()
+	logger := golog.NewTestLogger(t)
+
+	center := r3.Vector{X: 0, Y: 0, Z: 0}
+	side := 1.0
+
+	t.Run("Splitting empty octree node into octants", func(t *testing.T) {
+		basicOct, err := createNewOctree(ctx, center, side, logger)
+		test.That(t, err, test.ShouldBeNil)
+
+		err = basicOct.splitIntoOctants()
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, len(basicOct.node.tree), test.ShouldEqual, 8)
+		test.That(t, basicOct.node.nodeType, test.ShouldResemble, InternalNode)
+		test.That(t, basicOct.node.point, test.ShouldResemble, pc.PointAndData{})
+
+		for _, child := range basicOct.node.tree {
+			test.That(t, child.node.nodeType, test.ShouldResemble, LeafNodeEmpty)
+			test.That(t, child.node.point, test.ShouldResemble, pc.PointAndData{})
+		}
+	})
+
+	t.Run("Splitting filled basic octree node into octants", func(t *testing.T) {
+		basicOct, err := createNewOctree(ctx, center, side, logger)
+		test.That(t, err, test.ShouldBeNil)
+
+		p := r3.Vector{X: 0, Y: 0, Z: 0}
+		d := pc.NewValueData(1.0)
+
+		err = basicOct.Set(p, d)
+		test.That(t, err, test.ShouldBeNil)
+
+		err = basicOct.splitIntoOctants()
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, len(basicOct.node.tree), test.ShouldEqual, 8)
+		test.That(t, basicOct.node.nodeType, test.ShouldResemble, InternalNode)
+		test.That(t, basicOct.node.point, test.ShouldResemble, pc.PointAndData{})
+
+		filledLeaves := []*basicOctree{}
+
+		for _, child := range basicOct.node.tree {
+			if child.node.nodeType == LeafNodeFilled {
+				filledLeaves = append(filledLeaves, child)
+			}
+		}
+		test.That(t, len(filledLeaves), test.ShouldEqual, 1)
+		test.That(t, filledLeaves[0].node.point, test.ShouldResemble, pc.PointAndData{P: p, D: d})
+	})
+
+	t.Run("Splitting partially filled basic octree node into octants", func(t *testing.T) {
```

Comment:
[q] Should this test be called "Splitting internal basic octree node into octants"?